odoo/addons/base/models/qweb.py

```diff
@@ -655,115 +330,214 @@ def _compile_node(self, el, options):
+                result.append(' ' * space)
+            pos = t.start
+
+            if t.type == tokenize.ENCODING or t.type == token.ENDMARKER or t.type == token.DEDENT:
+                pass
+            elif t.type == token.OP and t.string in ['(', '[', '{']: # tuple, function, lambda, for
+                result.append(t.string)
+                brackets += 1
+                if index > 0 and t.string == '(' and tokens[index - 1].type == token.NAME:
+                    is_function_params.append(brackets)
+            elif t.type == token.OP and t.string in [')', ']', '}']: # tuple, function, lambda, for
+                result.append(t.string)
+                if is_function_params and is_function_params[-1] == brackets:
+                    is_function_params.pop()
+                while is_lambda_params and is_lambda_params[-1][0] >= brackets:
+                    is_lambda_params.pop()
+                brackets -= 1
+            elif t.type == token.OP and t.string == ',' and is_lambda_params and is_lambda_params[-1][0] >= brackets: # close lambda
+                while is_lambda_params and is_lambda_params[-1][0] >= brackets:
+                    is_lambda_params.pop()
+                result.append(t.string)
+            elif t.type == token.OP and t.string == '.': # attributes
+                result.append(t.string)
+                if index + 1 < tokens_len and tokens[index + 1].type == token.NAME:
+                    result.append(tokens[index + 1].string)
+                    index += 1
+            elif t.type == token.NAME and t.string == 'lambda': # lambda => allowed values
+                result.append('lambda ')
+                index += 1
+                keys = []
+                while tokens[index] and (tokens[index].type == token.OP or tokens[index].type == token.NAME) and tokens[index].string != ':':
+                    if tokens[index].type == token.NAME:
+                        key = tokens[index].string
+                        result.append(f'_qweb_temp_{key}')
+                        keys.append(key)
+                    else:
+                        result.append(tokens[index].string)
+                    index += 1
+                result.append(':')
+                is_lambda_params.append((brackets, keys))
+            elif t.type == token.NAME and t.string in allowed_keyword: # keywords
+                result.append(t.string)
+            elif t.type == token.NAME and any(t.string in p[1] for p in is_lambda_params): # allowed key word from lambda params
+                result.append(f'_qweb_temp_{t.string}')
+            elif t.type == token.NAME and any(p[0] <= brackets and t.string in p[1] for p in is_for_params): # allowed key word from lambda params
+                result.append(f'_qweb_temp_{t.string}')
+            elif t.type == token.NAME and is_function_params and index + 1 < tokens_len and tokens[index + 1].string == '=': # function kw
+                result.append(t.string)
```

Comment:
`is_function_params` is broken for anything but utterly trivial calls e.g. `(fn)(a=1)` will miscompile, so will `(lambda a: a)(a=1)`.

I'm not sure this level of complexity even makes sense? If we're within some sort of bracketing we can only ever have an expression, and assignments can't appear in such a context, so it I think it's always fine to emit the name as-is?

Reply:
The basic case is to change all 'a' values to values ['a']. The conditions added are each time to add support for one or the other case. It is not necessary to cover all the cases that may exist, but to cover our minimum needs. It is always possible for the dev to do multiple t-sets, or simply to send the correct values to the templates.